### DISCUSSION TITLE
Help for use WinStationShadow Win32 function

### DISCUSSION THREAD

**Discussion by szolnokit** [winsiderss/systeminformer#1124] [Questions / Answers] [answered]
I would like to use the WinStationShadow function in my small project. I would like to write a console RDP shadow switch utility. I found the Process Hacker use this function. See: https://github.com/processhacker/processhacker/blob/master/ProcessHacker/sessshad.c (at line 215)

My POC code:
```
using System;
using System.Runtime.InteropServices;

namespace ShadowSwitch
{
    class Program
    {
        public const short KBDSHIFT = 0x1;
        public const short KBDCTRL = 0x2;
        public const short KBDALT = 0x4;

        [DllImport(@"winsta.dll")]
        internal static extern Boolean WinStationShadow(IntPtr hServer, String TargetServerName, ulong TargetSessionId, byte HotKeyVk, ushort HotkeyModifiers);

        static void Main(string[] args)
        {
            Console.WriteLine(WinStationShadow(IntPtr.Zero, Environment.MachineName, uint.Parse(args[0]), (byte)'*', KBDCTRL | KBDALT));
            Console.WriteLine(Marshal.GetLastWin32Error());
        }
    }
}
```

But I get True as WinStationShadow result, and Win32 last error: 120 (This function is not supported on this system.)

I tested on same machine and with same (source/target) session and with same user account as like tested with Process Hacker. My shadow switching drop the error above, until Process Explorer shadow switch working as expected. 

What I wrong? Thank you for suggestion...

**Comment by dmex** [comment, marked as answer]
> True as WinStationShadow result

C# boolean is 4 bytes, C boolean is 1 byte. Your dllimport is missing `MarshalAs(UnmanagedType.I1)` so that result is incorrect. 

> Win32 last error: 120

Your dllimport is missing `SetLastError = true`. C# interop won't return the error code for the dllimport without it.

> (byte)'*'

Don't cast C# chars (UTF16) to C chars (ANSI).

> What I wrong?

Fix the type mashaling:

```
[DllImport("winsta.dll", CharSet = CharSet.Unicode, SetLastError = true)]
[return: MarshalAs(UnmanagedType.I1)]
public static extern Boolean WinStationShadow(IntPtr hServer, string serverName, uint targetSessionId, uint hotkeyVk, uint hotkeyModifier);
```

Use the correct types:

```
const uint KBDCTRL = 0x2;
const uint KBDALT = 0x4;
const uint VK_MULTIPLY = 0x6A;

Console.WriteLine(WinStationShadow(IntPtr.Zero, Environment.MachineName, 1, VK_MULTIPLY, KBDCTRL | KBDALT));
Console.WriteLine(Marshal.GetLastWin32Error());
```

**Reply by szolnokit** [reply]
Thank you for fast answer :) This was good at first try 👍
